libraries/plugins/bookie/bookie_plugin.cpp

```diff
@@ -58,6 +58,35 @@ namespace detail
  * they are created. 
  * We do this by creating a secondary index on bet_object.  We don't actually use it
  * to index any property of the bet, we just use it to register for callbacks.
+ *
+ * One objective of the helper is to watch "ephemeral" database objects
+ * in order to populate and persist its own own "persistent" objects within its own primary index.
+ *
+ * The life cycle of a naive helper object is:
+ *
+ * 1. During `plugin_initialize()`
+ *    The helper registers with the database for future notifications (Step 2) of the creation of a watched object
+ *    (`database().add_secondary_index<...>()`).
+ *    The helper also delegates the future persistence (Step 3) of its own primary index
+ *    to the database by registering it.
+ *    (`database().add_index<primary_index<...>>()`).
+ *    This primary index will be used to index "persistent" trackers of "ephemeral" objects that are observed during Step 2.
+ * 2. During a node session: the helper is notified by the database about the "ephemeral" watched objects
+ *    that are created during the node session through its `object_created()` callback.
+ *    During that callback, the helper will create a "persistent" version of the of "ephemeral" tracked object
+ *    and inject it into its own primary index.
+ * 3. When the node session shuts down, the database will persist the helper's primary index
+ *    that was registered during Step 1.
+ * 4. When the node is restarted, the database will automatically load the previous session(s) "persistent" objects
+ *    into the helper's primary index.
+ *
+ * The helper can ignore the `object_loaded()` events that are triggered when the watched "ephemeral" object type
+ * is loaded from persistence by the database because those objects were already processed
+ * by the the helper's `object_created()` during the prior sessions.
+ *
+ * NOTE: The helper should register itself for notifications of new "ephemeral" objects
+ *       during `plugin_initialize()` rather than `plugin_startup()`
+ *       for compatibility with a blockchain replay.
```

Comment:
Throughout this discussion, the database objects are referred to as "ephemeral," but I don't understand why... It would seem to imply that they are not persistent, but they are: this issue arose from the fact that they are loaded from disk at startup, just like the plugin's objects.

It would seem that the original idea of `secondary_index` was to be used with ephemeral objects, but bookie uses persistent ones which created the whole mess. Now we're extending `secondary_index` to allow for that use case.

Perhaps it would be better to talk about "node" objects, which belong to the main node logic, versus "plugin" objects, which belong to the plugin?